### PR TITLE
322945 add method to stop a batched job

### PR DESCRIPTION
**Pull request by Pespiri**

### CODE REVIEW

File: api/environments/job_handler.go [hunk 1]
```diff
@@ -71,6 +71,35 @@ func (eh EnvironmentHandler) GetJob(appName, envName, jobComponentName, jobName
...
+	if idx == -1 {
+		return jobNotFoundError(jobName)
+	}
+
+	jobStatus := slice.FindAll(batch.Status.JobStatuses, func(js radixv1.RadixBatchJobStatus) bool {
+		return js.Name == batchJobName && js.Phase != radixv1.BatchJobPhaseActive && js.Phase != radixv1.BatchJobPhaseWaiting && js.Phase != ""
+	})
+	if len(jobStatus) > 0 {
+		return radixhttp.ValidationError(jobName, fmt.Sprintf("invalid job running state=%s", jobStatus[0].Phase))
+	}
```
Comment (nilsgstrabo) [inline api/environments/job_handler.go:96]: A job without an entry in JobStatuses is considered to be waiting, and is stoppable. Consider moving the phase filter out of FindAll, and validate the Phase in line 94.

File: api/environments/job_handler.go [hunk 2]
```diff
@@ -137,6 +166,38 @@ func (eh EnvironmentHandler) getBatches(appName, envName, jobComponentName strin
...
 }
 
+// StopBatch Stop batch by name
+func (eh EnvironmentHandler) StopBatch(appName, envName, jobComponentName, batchName string) error {
+	batch, err := eh.getRadixBatch(appName, envName, jobComponentName, batchName, kube.RadixBatchTypeBatch)
+	if err != nil {
+		return err
+	}
+
+	if batch.Status.Condition.Type != radixv1.BatchConditionTypeActive && batch.Status.Condition.Type != radixv1.BatchConditionTypeWaiting {
```
Comment (nilsgstrabo) [inline api/environments/job_handler.go:176]: A RadixBatch that has not been processed by radix-operator will have an empty string ("") for Type. A user should be allowed to stop these batches

File: api/environments/job_handler.go [hunk 3]
```diff
@@ -137,6 +166,38 @@ func (eh EnvironmentHandler) getBatches(appName, envName, jobComponentName strin
...
+		return js.Phase != radixv1.BatchJobPhaseActive && js.Phase != radixv1.BatchJobPhaseWaiting && js.Phase != ""
+	})
+
+	didChange := false
+	for idx, job := range batch.Spec.Jobs {
+		if slice.FindIndex(jobStatus, func(js radixv1.RadixBatchJobStatus) bool { return js.Name == job.Name }) == -1 {
+			batch.Spec.Jobs[idx].Stop = radixutils.BoolPtr(true)
+			didChange = true
+		}
+	}
```
Comment (nilsgstrabo) [inline api/environments/job_handler.go:191]: Create common function/logic to check if a job is stoppable, and use this in both StopBatch and StopJob

**Review by nilsgstrabo — CHANGES_REQUESTED** [review]
See comments

File: api/environments/job_handler.go (hunk 1 above)
Comment (Pespiri) [reply api/environments/job_handler.go:96]: the no status is already accounted for, but the validation will be rewritten to be easier to read

File: api/environments/environment_controller.go [hunk 4]
```diff
@@ -1246,6 +1256,67 @@ func GetJob(accounts models.Accounts, w http.ResponseWriter, r *http.Request) {
...
+	//   required: false
+	// - name: Impersonate-Group
+	//   in: header
+	//   description: Works only with custom setup of cluster. Allow impersonation of test group (Required if Impersonate-User is set)
+	//   type: string
+	//   required: false
+	// responses:
+	//   "204":
+	//     description: "Success"
+	//   "401":
```
[the remark this answers is not in the document]
Comment (Pespiri) [reply api/environments/environment_controller.go:1300]: fixed

File: api/environments/job_handler.go (hunk 1 above)
Comment (Pespiri) [reply api/environments/job_handler.go]: fixed

File: api/environments/job_handler.go (hunk 2 above)
Comment (Pespiri) [reply api/environments/job_handler.go]: fixed

File: api/environments/job_handler.go (hunk 3 above)
Comment (Pespiri) [reply api/environments/job_handler.go]: fixed

File: api/environments/environment_controller_test.go [hunk 5]
```diff
@@ -2049,6 +2049,224 @@ func Test_GetBatches_JobListShouldBeEmpty(t *testing.T) {
...
+			Status: v1.RadixBatchStatus{
+				JobStatuses: []v1.RadixBatchJobStatus{
+					validJobs[0].jobStatus, validJobs[1].jobStatus, validJobs[2].jobStatus, validJobs[3].jobStatus,
+					invalidJobs[0].jobStatus, invalidJobs[1].jobStatus, invalidJobs[2].jobStatus,
+				},
+			},
+		},
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:   batch2Name,
```
[the remark this answers is not in the document]
Comment (Pespiri) [reply api/environments/environment_controller_test.go]: rewrote tests for better code coverage

**Review by nilsgstrabo — APPROVED** [review]

**Pull request merged**